cmd/admin/cert.go

```diff
@@ -0,0 +1,260 @@
+	// selection flag was given on the command line.
+	setInputs := map[string]bool{
+		"-serial":         *serial != "",
+		"-incident-table": *incidentTable != "",
+		"-serials-file":   *serialsFile != "",
+		"-private-key":    *privKey != "",
+		"-reg-id":         *regID != 0,
+	}
+	maps.DeleteFunc(setInputs, func(_ string, v bool) bool { return !v })
+	if len(setInputs) == 0 {
+		return errors.New("at least one input method flag must be specified")
+	} else if len(setInputs) > 1 {
+		return fmt.Errorf("more than one input method flag specified: %v", maps.Keys(setInputs))
+	}
+
+	var serials []string
+	var err error
+	switch maps.Keys(setInputs)[0] {
+	case "-serial":
+		serials, err = []string{*serial}, nil
+	case "-incident-table":
+		serials, err = a.serialsFromIncidentTable(ctx, *incidentTable)
+	case "-serials-file":
+		serials, err = a.serialsFromFile(ctx, *serialsFile)
+	case "-private-key":
+		serials, err = a.serialsFromPrivateKey(ctx, *privKey)
+	case "-reg-id":
+		serials, err = a.serialsFromRegID(ctx, int64(*regID))
+	default:
+		return errors.New("no recognized input method flag set (this shouldn't happen)")
+	}
+	if err != nil {
+		return fmt.Errorf("collecting serials to revoke: %w", err)
+	}
+
+	if len(serials) == 0 {
+		return errors.New("no serials to revoke found")
+	}
+	a.log.Infof("Found %d certificates to revoke", len(serials))
+
+	err = a.revokeSerials(ctx, serials, revocation.Reason(*reason), *malformed, *skipBlock, int(*parallelism))
+	if err != nil {
+		return fmt.Errorf("revoking serials: %w", err)
+	}
+
+	return nil
+}
+
+func (a *admin) serialsFromIncidentTable(ctx context.Context, tableName string) ([]string, error) {
+	stream, err := a.sac.SerialsForIncident(ctx, &sapb.SerialsForIncidentRequest{IncidentTable: tableName})
+	if err != nil {
+		return nil, fmt.Errorf("setting up stream of serials from incident table %q: %s", tableName, err)
+	}
+
+	var serials []string
+	for {
+		is, err := stream.Recv()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, fmt.Errorf("streaming serials from incident table %q: %s", tableName, err)
+		}
+		serials = append(serials, is.Serial)
+	}
+
+	return serials, nil
+}
+
+func (a *admin) serialsFromFile(_ context.Context, filePath string) ([]string, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("opening serials file: %w", err)
+	}
+
+	var serials []string
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		serial := scanner.Text()
+		if serial == "" {
+			continue
+		}
+		serials = append(serials, serial)
+	}
+
+	return serials, nil
+}
+
+func (a *admin) serialsFromPrivateKey(ctx context.Context, privkeyFile string) ([]string, error) {
+	_, publicKey, err := privatekey.Load(privkeyFile)
+	if err != nil {
+		return nil, fmt.Errorf("loading private key file: %w", err)
+	}
+
+	spkiHash, err := core.KeyDigest(publicKey)
+	if err != nil {
+		return nil, fmt.Errorf("computing SPKI hash: %w", err)
+	}
+
```

Comment:
Lines 172-180 are a duplicate of  `a.spkiHashFromPrivateKey` in `//cmd/admin/key.go`. These lines could become
```
spkiHash, err := a.spkiHashFromPrivateKey(privkeyFile)
if err != nil {
	return err
}
```